### PR TITLE
Parsing of the metadata response (RFC5464)

### PR DESCRIPTION
**Pull request by Sergey-Kitov**
Implementation of METADATA extension response parser

### CODE REVIEW

**Comment by djc** [comment]
Thanks for working on this!

Functionally this looks pretty okay. Would you mind reformatting it a bunch so that the style within the macro calls looks more like rustfmt-formatted code? Notably:

* Opening braces for `if` conditions go on the same line
* Max line length at 100 characters
* Closing brace for multi-line blocks on its own line (possibly with comma, but not at the last line of the block
* Spaces inside braces for structs on the same line

**Comment by Sergey-Kitov** [comment]
Does it look better now?

File: imap-proto/src/parser/rfc3501.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@ use nom::IResult;
 use std::str;
 
 use crate::parser::rfc4551;
+use crate::parser::rfc5464::*;
```
Comment (djc) [inline imap-proto/src/parser/rfc3501.rs]: Okay, please don't use `*` imports -- sorry, should have mentioned that the first time around.

File: imap-proto/src/parser/rfc3501.rs [hunk 2]
```diff
@@ -47,7 +48,7 @@ named!(status<Status>, alt!(
     status_bye
 ));
 
-named!(mailbox<&str>, map!(
+named!(pub(crate) mailbox<&str>, map!(
```
Comment (djc) [inline imap-proto/src/parser/rfc3501.rs]: Why make this `pub(crate)`? That seems unrelated to the main goal of this PR.

File: imap-proto/src/parser/rfc5464.rs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    VendorComment(usize),
+    Path(usize),
+    Done(usize),
+    Fail(nom::Err<&'a [u8]>),
+}
+
+fn check_private_shared(i: &[u8]) -> EntryParseStage {
+    if i.starts_with(b"/private") {
+        return EntryParseStage::VendorComment(8);
+    }
```
Comment (djc) [inline imap-proto/src/parser/rfc5464.rs]: Using `else if` would be preferable here (saves one line and logically makes sense).

File: imap-proto/src/parser/rfc5464.rs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    if i.starts_with(b"/shared") {
+        return EntryParseStage::Admin(7);
+    }
+    return EntryParseStage::Fail(
+        nom::Err::Error(
+        b"Entry Name doesn't start with /private or /shared"));
```
Comment (djc) [inline imap-proto/src/parser/rfc5464.rs]: Please put the closing parenthesis on the next line here.

File: imap-proto/src/parser/rfc5464.rs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+fn check_admin(i: &[u8], l:usize) -> EntryParseStage {
+    if i[l..].starts_with(b"/admin") {
+        return EntryParseStage::Path(l+6);
```
Comment (djc) [inline imap-proto/src/parser/rfc5464.rs]: Please put spaces around the `+` here.

**Review by djc — COMMENTED** [review]
There's a bunch of formatting issues still. I would probably enable formatting in the new module, format all the non-macro code and then manually format the macro code again (or find some other way to properly format the non-macro code).

File: imap-proto/src/parser/rfc3501.rs (hunk 2 above)
Comment (Sergey-Kitov) [reply imap-proto/src/parser/rfc3501.rs]: There was some reason for making it pub(crate), but it looks like the reason has gone during development.

File: imap-proto/src/parser/rfc5464.rs (hunk 4 above)
Comment (Sergey-Kitov) [reply imap-proto/src/parser/rfc5464.rs]: May be it is better to put error message on the same line as  "...Error("

Comment (djc) [reply imap-proto/src/parser/rfc5464.rs]: Please do whatever `rustfmt` does in this case.

**Comment by Sergey-Kitov** [comment]
I have implemented all rustfmt corrections for non macro invocation code.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Thanks!

**Comment by Sergey-Kitov** [comment]
Thank you very much!
